namesys/pubsub.go

```diff
@@ -0,0 +1,346 @@
+	return resolve(ctx, r, name, depth, "/ipns/")
+}
+
+func (r *PubsubResolver) resolveOnce(ctx context.Context, name string) (value path.Path, err error) {
+	log.Debugf("PubsubResolve: '%s'", name)
+
+	// retrieve the public key once (for verifying messages)
+	xname := strings.TrimPrefix(name, "/ipns/")
+	hash, err := mh.FromB58String(xname)
+	if err != nil {
+		log.Warningf("PubsubResolve: bad input hash: [%s]", xname)
+		return
+	}
+
+	pubk, err := r.pkf.GetPublicKey(ctx, peer.ID(hash))
+	if err != nil {
+		log.Warningf("PubsubResolve: error fetching public key: %s [%s]", err.Error(), xname)
+		return
+	}
+
+	// the mutex is locked both for subscription map manipulation and datastore
+	// manipulation (otherwise race in TTL checking/invalidation)
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	// see if we already have a pubsub subscription; if not, subscribe
+	sub, ok := r.subs[name]
+	if !ok {
+		sub, err = r.ps.Subscribe(name)
+		if err != nil {
+			return
+		}
+
+		r.subs[name] = sub
+		go r.handleSubscription(sub, name, pubk)
+		go bootstrapPubsub(r.ctx, r.cr, r.host, name)
+	}
+
+	// resolve to what we may already have in the datastore
+	dsval, err := r.ds.Get(dshelp.NewKeyFromBinary([]byte(name)))
+	if err != nil {
+		// this signals ds.ErrNotFound when we have no record
+		return
+	}
+
+	entry := dsval.(*pb.IpnsEntry)
+
+	// check EOL; if the entry has expired, delete from datastore and return ds.ErrNotFound
+	eol, ok := checkEOL(entry)
+	if ok && eol.Before(time.Now()) {
+		r.ds.Delete(dshelp.NewKeyFromBinary([]byte(name)))
+		return "", ds.ErrNotFound
+	}
+
+	value, err = path.ParsePath(string(entry.GetValue()))
+	return
+}
+
+func (r *PubsubResolver) handleSubscription(sub *floodsub.Subscription, name string, pubk ci.PubKey) {
+	defer sub.Cancel()
+
+	for {
+		msg, err := sub.Next(r.ctx)
+		if err != nil {
+			if err != context.Canceled {
+				log.Warningf("PubsubResolve: subscription error in %s: %s", name, err.Error())
+			}
+			return
+		}
+
+		err = r.receive(msg, name, pubk)
+		if err != nil {
+			log.Warningf("PubsubResolve: error proessing update for %s: %s", name, err.Error())
+		}
+	}
+}
+
+func (r *PubsubResolver) receive(msg *floodsub.Message, name string, pubk ci.PubKey) error {
+	data := msg.GetData()
+	if data == nil {
+		return errors.New("empty message")
+	}
+
+	entry := new(pb.IpnsEntry)
+	err := proto.Unmarshal(data, entry)
+	if err != nil {
+		return err
+	}
+
+	ok, err := pubk.Verify(ipnsEntryDataForSig(entry), entry.GetSignature())
+	if err != nil || !ok {
+		return errors.New("signature verification failed")
+	}
+
+	_, err = path.ParsePath(string(entry.GetValue()))
+	if err != nil {
+		return err
+	}
+
+	eol, ok := checkEOL(entry)
+	if ok && eol.Before(time.Now()) {
+		return errors.New("stale update; EOL exceeded")
+	}
+
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	// check the sequence number against what we may already have in our datastore
+	dsval, err := r.ds.Get(dshelp.NewKeyFromBinary([]byte(name)))
```

Comment:
serialization, this will come back as bytes (just doesnt happen to with the in memory datastore)